### PR TITLE
Add `join_paths` function to correctly join file paths

### PR DESCRIPTION
**Pull request by jdidion**
### Checklist
- [x] Pull request details were added to CHANGELOG.md
- [x] Valid examples WDL's were added or updated to the SPEC.md (see the [guide](https://github.com/openwdl/wdl-tests/blob/main/docs/MarkdownTests.md) on writing markdown tests)

### CODE REVIEW

File: SPEC.md [hunk 1]
```diff
@@ -7367,6 +7368,93 @@ Example output:
 </p>
 </details>
 
+### ✨ `join_paths`
+
+```
+File join_paths(File|String, String)
```
Comment (markjschreiber) [inline SPEC.md]: This probably depends on the rules of casting between `File` and `Directory` but should the definitions be

```
File join_paths(Directory|String, String)
```
etc

**Review by markjschreiber — COMMENTED** [review]
Question about `File` vs `Directory` in the function definition.

File: SPEC.md (hunk 1 above)
Comment (jdidion) [reply SPEC.md]: We haven't added `Directory` support to 1.2 yet :)

**Review by markjschreiber — APPROVED** [review]
LGTM. One recommendation that we could make is that engines should normalize the final path produced by `join_paths` because when a workflow output location is specified by using `join_paths` and the output will be going to `s3` or `gs` or similar object stores then `foo//baa/./baz` is not the same as `foo/baa/baz` even though they are when the output location is a POSIX compliant FS.

File: SPEC.md [hunk 2]
```diff
@@ -7367,6 +7368,93 @@ Example output:
...
+File join_paths(File, Array[String]+)
+File join_paths(Array[String]+)
+```
+
+Joins together two or more paths into an absolute path in the host filesystem.
+
+There are three variants of this function:
+
+1. `File join_paths(File|String, String)`: Joins together exactly two paths. The first path may be either absolute or relative and must specify a directory; the second path is relative to the first path and may specify a file or directory.
```
Comment (illusional) [inline SPEC.md]: If File is coercible to a String, is this needed, or just helpful to imply?

File: SPEC.md [hunk 3]
```diff
@@ -7367,6 +7368,93 @@ Example output:
...
+2. `File join_paths(File, Array[String]+)`: Joins together any number of relative paths with a base path. The first argument may be either an absolute or a relative path and must specify a directory. The paths in the second array argument must all be relative. The *last* element may specify a file or directory; all other elements must specify a directory.
+3. `File join_paths(Array[String]+)`: Joins together any number of paths. The array must not be empty. The *first* element of the array may be either absolute or relative; subsequent path(s) must be relative. The *last* element may specify a file or directory; all other elements must specify a directory.
+
+An absolute path starts with `/` and indicates that the path is relative to the root of the container in which the task is executed. A relative path does not start with `/` and indicates the path is relative to its parent directory. It is up to the execution engine to determine which directory to use as the parent when resolving relative paths; by default it is the working directory in which the task is executed.
```
Comment (illusional) [inline SPEC.md]: Do we want to prescribe the behaviour of supplying an absolute path as a later parameter, eg: Python will ignore all previous paths - or should this be an exception?

File: SPEC.md [hunk 4]
```diff
@@ -7367,6 +7368,93 @@ Example output:
...
+    File rel_dir_file = "mydir"
+    String rel_str = "mydata.txt"
+  }
+
+  # these are all equivalent to '/usr/bin/echo'
+  File bin1 = join_paths(abs_file, [rel_str_dir, rel_file])
+  File bin2 = join_paths(abs_str, [rel_str_dir, rel_file])
+  File bin3 = join_paths([abs_str, rel_str_dir, rel_file])
+  
+  # this resolves to '<working dir>/mydir/mydata.txt'
```
Comment (illusional) [inline SPEC.md]: Depends how pedantic the spec should be, but the above states:

> It is up to the execution engine to determine which directory to use as the parent when resolving relative paths;

So maybe "the default behaviour is this should resolve to ..."?

File: SPEC.md (hunk 2 above)
Comment (jdidion) [reply SPEC.md]: Good point. `String` is coercible to `File` so we only need `File`.

File: SPEC.md (hunk 3 above)
Comment (jdidion) [reply SPEC.md]: I don't really like that behavior in python tbh. IMO if we're going to the principle of least surprising behavior, then that should be an error.

**Review by illusional — APPROVED** [review]
Awesome!! Thanks for driving this :)

**Pull request merged**